File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode498DiagonalTraverse.cpp
#include "LeetCode498DiagonalTraverse.h"

#include "Auxilaries.h"
using namespace std;

vector<int> LeetCode498DiagonalTraverse::findDiagonalOrder(vector<vector<int>>& mat)
{
    int heigth = static_cast<int>(mat.size());
    int width = static_cast<int>(mat[0].size());
    pair<int, int> startPos = {0, 0};
    int slope = 1;
    vector<int> result;
    result.reserve(heigth * width);
    pair<int, int> currentPos;

    while (result.size() < heigth * width)
    {
        currentPos = make_pair(startPos.first, startPos.second);
        if (slope == 1)
        {
            while (currentPos.first>=0&&currentPos.second<width)
            {
                result.push_back(mat[currentPos.first][currentPos.second]);
                currentPos.first--;
                currentPos.second++;
            }
            currentPos.second --;
            currentPos.first++;
            if (currentPos.second<width-1)
            {
                startPos.second=currentPos.second+1;
                startPos.first = currentPos.first;
            }
            else
            {
                startPos.first=currentPos.first+1;
                startPos.second = currentPos.second;
            }
        }
        else
        {
            while (currentPos.second>=0&&currentPos.first<heigth)
            {
                result.push_back(mat[currentPos.first][currentPos.second]);
                currentPos.first++;
                currentPos.second--;
            }
            currentPos.first--;
            currentPos.second ++;
            if (currentPos.first<heigth-1)
            {
                startPos.first=currentPos.first+1;
                startPos.second = currentPos.second;
            }
            else
            {
                startPos.second=currentPos.second+1;
                startPos.first = currentPos.first;
            }
        }
        slope *= -1;
    }
    return result;
}

void LeetCode498DiagonalTraverse::Run()
{
    vector<vector<int>> mat = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    vector<int> result = findDiagonalOrder(mat);
    Auxilaries::PrintVector(result);
}

File: LeetCodeProblemsGo/LeetCode498DiagonalTraverse/LeetCode498DiagonalTraverse.go
package LeetCode498DiagonalTraverse

func findDiagonalOrder(mat [][]int) []int {
	heigth := len(mat)
	width := len(mat[0])
	startPos := []int{0, 0}
	currentPos := []int{0, 0}
	isGoingDown := true
	result := make([]int, 0, heigth*width)
	for len(result) < heigth*width {
		currentPos[0], currentPos[1] = startPos[0], startPos[1]
		if isGoingDown {
			for currentPos[0] >= 0 && currentPos[1] < width {
				result = append(result, mat[currentPos[0]][currentPos[1]])
				currentPos[0]--
				currentPos[1]++
			}
			currentPos[0]++
			currentPos[1]--
			if currentPos[1]+1 < width {
				startPos[1] = currentPos[1] + 1
				startPos[0] = currentPos[0]
			} else {
				startPos[0] = currentPos[0] + 1
				startPos[1] = currentPos[1]
			}
		} else {
			for currentPos[1] >= 0 && currentPos[0] < heigth {
				result = append(result, mat[currentPos[0]][currentPos[1]])
				currentPos[0]++
				currentPos[1]--
			}
			currentPos[0]--
			currentPos[1]++
			if currentPos[0]+1 < heigth {
				startPos[0] = currentPos[0] + 1
				startPos[1] = currentPos[1]
			} else {
				startPos[1] = currentPos[1] + 1
				startPos[0] = currentPos[0]
			}
		}
		isGoingDown = !isGoingDown
	}
	return result
}
